### PR TITLE
Add citadel to new test framework

### PR DESCRIPTION
**Pull request by JimmyCYJ**
Support new test framework to test Citadel in Kubernetes. 
Add a test to verify basic behavior of Citadel.

### CODE REVIEW

File: tests/integration2/citadel/citadel_test.go [hunk 1]
```diff
@@ -0,0 +1,58 @@
...
+
+// TestSecretCreationKubernetes verifies that Citadel creates secret and stores as Kubernetes secrets,
+// and that when secrets are deleted, new secrets will be created.
+func TestSecretCreationKubernetes(t *testing.T) {
+	framework.Requires(t, dependency.Kubernetes, dependency.Citadel)
+	env := framework.AcquireEnvironment(t)
+	c := env.GetCitadelOrFail(t)
+
+	// Test the existence of istio.default secret.
+	s, err := c.WaitForSecretExist()
```
[the remark this answers is not in the document]
Comment (hklai) [reply tests/integration2/citadel/citadel_test.go]: +1

Then WaitForSecretExist should become CheckSecretExists or something along that line,

File: pkg/test/framework/environment.go [hunk 2]
```diff
@@ -111,6 +111,31 @@ func (e *environment) GetPilotOrFail(t testing.TB) env.DeployedPilot {
...
+
+	m, err := e.GetCitadel()
+	if err != nil {
+		t.Fatal(err)
+	}
+
+	return m
+}
+
+func (e *environment) CitadelName() string {
```
[the remark this answers is not in the document]
Comment (nmittler) [reply pkg/test/framework/environment.go]: +1 ... I think you should remove this.

File: pkg/test/framework/environment/environment.go [hunk 3]
```diff
@@ -214,6 +222,14 @@ type (
 		// The raw content of the response
 		Raw string
 	}
+
+	// DeployedCitadel represents a deployed Citadel instance.
+	DeployedCitadel interface {
```
Comment (nmittler) [inline pkg/test/framework/environment/environment.go]: I believe the lifecycle of the secret should be tied to the lifecycle of this component (e.g. the component doesn't initialize until the secret exists), no? If so, I don't think you need these methods at all.

File: tests/integration2/citadel/citadel_test.go (hunk 1 above)
[the remark this answers is not in the document]
Comment (JimmyCYJ) [reply tests/integration2/citadel/citadel_test.go]: In this test, the returned error indicates test failure. I am going to migrate TestCertificateRotation in the next PR, which waits for new secrets and uses returned error as a signal to retry.

File: pkg/test/framework/environment/environment.go [hunk 4]
```diff
@@ -214,6 +222,14 @@ type (
...
+		Deployed
+
+		WaitForSecretExist() (*v1.Secret, error)
```
[the remark this answers is not in the document]
Comment (JimmyCYJ) [reply pkg/test/framework/environment/environment.go]: This test is migrated from [TestSecretCreation](https://github.com/istio/istio/blob/master/security/tests/integration/secretCreationTest/secret_creation_test.go#L39). I have renamed to method to WaitForSecretToExist(). Thanks.

File: pkg/test/framework/environment.go (hunk 2 above)
[the remark this answers is not in the document]
Comment (JimmyCYJ) [reply pkg/test/framework/environment.go]: It should be removed. Done.

File: pkg/test/framework/components/citadel/citadel.go [hunk 5]
```diff
@@ -0,0 +1,114 @@
...
+	"istio.io/istio/pkg/test/framework/dependency"
+	"istio.io/istio/pkg/test/framework/environment"
+	"istio.io/istio/pkg/test/framework/environments/kubernetes"
+)
+
+const (
+	// Specifies how long we wait before a secret becomes existent.
+	secretWaitTime = 20 * time.Second
+	// Name of secret created by Citadel
+	secretName = "istio.default"
```
[the remark this answers is not in the document]
Comment (JimmyCYJ) [reply pkg/test/framework/components/citadel/citadel.go]: Yes, it is always "istio.default" in old security integration tests. Maybe we can keep it unless we have more tests which require other secret names.

File: pkg/test/framework/environment/environment.go (hunk 3 above)
Comment (JimmyCYJ) [reply pkg/test/framework/environment/environment.go]: The lifecycle might not be the same as lifecycle of Citadel. I am going to migrate another integrate test [TestCertificateRotation](https://github.com/istio/istio/blob/19315ed01763585525377460a7fa72237dc3107c/security/tests/integration/certificateRotationTest/certificate_rotation_test.go#L45) which configures Citadel to 
generate short life secrets [here](https://github.com/istio/istio/blob/19315ed01763585525377460a7fa72237dc3107c/security/tests/integration/istio_ca_cert_rotation_env.go#L87).

**Comment by JimmyCYJ** [comment]
Added Citadel author @myidpt

**Comment by hklai** [comment]
Looks like the unit test jobs are failing even all tests have passed....

**Comment by JimmyCYJ** [comment]
The new test starts to fail today, I didn't make any change to this PR on citadel part. Perhaps there is an update to test framework that causes the failure. @ozevren would you mind to take a look?
```unable to recognize "/tmp/citadel-test-edf8b2e88e2c43969e/istio-system-1538091390194412483.yaml": no matches for networking.istio.io/, Kind=Gateway
unable to recognize "/tmp/citadel-test-edf8b2e88e2c43969e/istio-system-1538091390194412483.yaml": no matches for config.istio.io/, Kind=attributemanifest
unable to recognize "/tmp/citadel-test-edf8b2e88e2c43969e/istio-system-1538091390194412483.yaml": no matches for config.istio.io/, Kind=attributemanifest
unable to recognize "/tmp/citadel-test-edf8b2e88e2c43969e/istio-system-1538091390194412483.yaml": no matches for config.istio.io/, Kind=handler
unable to recognize "/tmp/citadel-test-edf8b2e88e2c43969e/istio-system-1538091390194412483.yaml": no matches for config.istio.io/, Kind=logentry
unable to recognize "/tmp/citadel-test-edf8b2e88e2c43969e/istio-system-1538091390194412483.yaml": no matches for config.istio.io/, Kind=logentry
unable to recognize "/tmp/citadel-test-edf8b2e88e2c43969e/istio-system-1538091390194412483.yaml": no matches for config.istio.io/, Kind=rule
unable to recognize "/tmp/citadel-test-edf8b2e88e2c43969e/istio-system-1538091390194412483.yaml": no matches for config.istio.io/, Kind=rule
unable to recognize "/tmp/citadel-test-edf8b2e88e2c43969e/istio-system-1538091390194412483.yaml": no matches for config.istio.io/, Kind=metric
unable to recognize "/tmp/citadel-test-edf8b2e88e2c43969e/istio-system-1538091390194412483.yaml": no matches for config.istio.io/, Kind=metric
unable to recognize "/tmp/citadel-test-edf8b2e88e2c43969e/istio-system-1538091390194412483.yaml": no matches for config.istio.io/, Kind=metric
unable to recognize "/tmp/citadel-test-edf8b2e88e2c43969e/istio-system-1538091390194412483.yaml": no matches for config.istio.io/, Kind=metric
unable to recognize "/tmp/citadel-test-edf8b2e88e2c43969e/istio-system-1538091390194412483.yaml": no matches for config.istio.io/, Kind=metric
unable to recognize "/tmp/citadel-test-edf8b2e88e2c43969e/istio-system-1538091390194412483.yaml": no matches for config.istio.io/, Kind=metric
unable to recognize "/tmp/citadel-test-edf8b2e88e2c43969e/istio-system-1538091390194412483.yaml": no matches for config.istio.io/, Kind=handler
unable to recognize "/tmp/citadel-test-edf8b2e88e2c43969e/istio-system-1538091390194412483.yaml": no matches for config.istio.io/, Kind=rule
unable to recognize "/tmp/citadel-test-edf8b2e88e2c43969e/istio-system-1538091390194412483.yaml": no matches for config.istio.io/, Kind=rule
unable to recognize "/tmp/citadel-test-edf8b2e88e2c43969e/istio-system-1538091390194412483.yaml": no matches for config.istio.io/, Kind=handler
unable to recognize "/tmp/citadel-test-edf8b2e88e2c43969e/istio-system-1538091390194412483.yaml": no matches for config.istio.io/, Kind=rule
unable to recognize "/tmp/citadel-test-edf8b2e88e2c43969e/istio-system-1538091390194412483.yaml": no matches for config.istio.io/, Kind=rule
unable to recognize "/tmp/citadel-test-edf8b2e88e2c43969e/istio-system-1538091390194412483.yaml": no matches for config.istio.io/, Kind=kubernetes
unable to recognize "/tmp/citadel-test-edf8b2e88e2c43969e/istio-system-1538091390194412483.yaml": no matches for networking.istio.io/, Kind=DestinationRule
unable to recognize "/tmp/citadel-test-edf8b2e88e2c43969e/istio-system-1538091390194412483.yaml": no matches for networking.istio.io/, Kind=DestinationRule

2018-09-27T23:36:32.876356Z	info	CI	=== FAILED: Deploy Istio ===
```

File: tests/integration2/citadel/secret_creation_test.go [hunk 6]
```diff
@@ -0,0 +1,58 @@
...
+		t.Error(err)
+	}
+
+	t.Log(`secret "istio.default" has been deleted`)
+
+	// Test that the deleted secret is re-created properly.
+	if _, err := c.WaitForSecretToExist(); err != nil {
+		t.Error(err)
+	}
+	t.Log(`checking secret "istio.default" is correctly re-created`)
```
Comment (myidpt) [inline tests/integration2/citadel/secret_creation_test.go]: Call ExamineSecret(s)?

**Comment by JimmyCYJ** [comment]
TestIngessToPrometheus_ServiceMetric - mixer failed. @ozevren, would you mind to take a look?
[link](https://circleci.com/gh/istio/istio/205189?utm_campaign=vcs-integration-link&utm_medium=referral&utm_source=github-build-link)

**Comment by ozevren** [comment]
That is a known issue. I'll follow up on this soon.

**Comment by hklai** [comment]
I have seen that (all tests passed but job failed) before, but it went away. Did you recently rebase?

**Comment by JimmyCYJ** [comment]
> I have seen that (all tests passed but job failed) before, but it went away. Did you recently rebase?

I just use git merge upstream/master again, and the unit test jobs still fail with zero test failures.

**Comment by hklai** [comment]
I notice "make test common-tests" failed. I ran it locally and noticed two format related errors.
Sent https://github.com/istio/istio/pull/9110 to fix them, and will check if unit tests pass there... I cannot explain why this happens to only your PR but not the others.

**Comment by hklai** [comment]
Now that https://github.com/istio/istio/pull/9110 is merged, could you rebase to it and try again?

**Comment by hklai** [comment]
OK it looks like I fixed the tests with https://github.com/istio/istio/pull/8932/commits/6ba23f949cac60398cce96539dfc9ffc9e5bbffe =)

Yeah we prob should fix make test to spit out better error messages.

**Comment by JimmyCYJ** [comment]
@hklai Thanks a lot, Andy!

**Pull request merged**